### PR TITLE
[CARBONDATA-4143] Enable UT with index server and fix related issues

### PR DESCRIPTION
**Pull request by ShreelekhyaG**
### Why is this PR needed?
enable to run UT with index server.
Fix below issues:
1. With index server enabled, select query gives incorrect result with SI when parent and child table segments are not in sync.
2. When reindex is triggered, if stale files are present in the segment directory the segment file is being written with incorrect file names. (both valid index and stale mergeindex file names). As a result, duplicate data is present in SI table but there are no error/incorrect query results.
 
 ### What changes were proposed in this PR?
usage of flag `useIndexServer`. excluded some of the test cases to not run with index server.
1. While pruning from index server, `missingSISegments` values were not getting considered. Have passed down and set those values to filter.
2. Before loading data to SI segment, added changes to delete the segment directory if already present.
    
 ### Does this PR introduce any user interface change?
 - No

 ### Is any new testcase added?
 - Yes

### CODE REVIEW

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12444/job/ApacheCarbon_PR_Builder_2.4.5/3761/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12444/job/ApacheCarbonPRBuilder2.3/5526/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/3338/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/5097/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/3352/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/5111/

**Comment by ShreelekhyaG** [comment]
retest this please

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/3355/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/5114/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12444/job/ApacheCarbon_PR_Builder_2.4.5/3765/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12444/job/ApacheCarbonPRBuilder2.3/5530/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/3292/

**Comment by ydvpankaj99** [comment]
retest this please

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/5050/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12444/job/ApacheCarbonPRBuilder2.3/5539/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12444/job/ApacheCarbon_PR_Builder_2.4.5/3776/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/5062/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_2.4.5/3310/

**Comment by brijoobopanna** [comment]
retest this please

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_2.4.5/3611/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12602/job/ApacheCarbonPRBuilder2.3/5356/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_2.4.5/3637/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12602/job/ApacheCarbonPRBuilder2.3/5382/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12602/job/ApacheCarbonPRBuilder2.3/5411/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_2.4.5/3691/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12602/job/ApacheCarbonPRBuilder2.3/5436/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12602/job/ApacheCarbonPRBuilder2.3/5448/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_2.4.5/3704/

File: integration/spark/src/main/scala/org/apache/spark/sql/test/SparkTestQueryExecutor.scala [hunk 1]
```diff
@@ -82,6 +83,15 @@ object SparkTestQueryExecutor {
       copyResourcesifNotExists(hdfsUrl, s"$integrationPath/spark/src/test/resources",
         s"$integrationPath//spark-common-cluster-test/src/test/resources/testdatafileslist.txt")
   }
+  if (System.getProperty("useIndexServer") != null || System.getenv("useIndexServer") != null) {
```
Comment (Indhumathi27) [inline integration/spark/src/main/scala/org/apache/spark/sql/test/SparkTestQueryExecutor.scala:86]: Please check and keep single check to useIndexServer

File: integration/spark/src/main/scala/org/apache/spark/sql/test/util/CarbonFunSuite.scala [hunk 2]
```diff
@@ -44,4 +44,18 @@ private[spark] abstract class CarbonFunSuite extends FunSuite {
     }
   }
 
+  protected def test(testName : scala.Predef.String, ignoreForIndexServer: Boolean,
+                     testTags : org.scalatest.Tag*)
+                    (testFun : => Unit): Unit = {
+    if (ignoreForIndexServer) {
+      if (System.getProperty("useIndexServer") == null) {
```
Comment (Indhumathi27) [inline integration/spark/src/main/scala/org/apache/spark/sql/test/util/CarbonFunSuite.scala:51]: Please add the comment

File: integration/spark/src/main/scala/org/apache/carbondata/indexserver/IndexServer.scala [hunk 3]
```diff
@@ -50,6 +51,7 @@ trait ServerInterface {
   /**
    * Used to prune and cache the index for the table.
    */
+  @throws(classOf[IOException])
```
Comment (Indhumathi27) [inline integration/spark/src/main/scala/org/apache/carbondata/indexserver/IndexServer.scala:54]: can ignore the testcase and revert this change

**Comment by Indhumathi27** [comment]
@ShreelekhyaG Please attach the CI snapshot when run with indexserver enabled

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_2.4.5/3734/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12602/job/ApacheCarbonPRBuilder2.3/5478/

**Comment by ShreelekhyaG** [comment]
retest this please

File: integration/spark/src/main/scala/org/apache/spark/sql/test/SparkTestQueryExecutor.scala (hunk 1 above)
Comment (ShreelekhyaG) [reply integration/spark/src/main/scala/org/apache/spark/sql/test/SparkTestQueryExecutor.scala]: ok

File: integration/spark/src/main/scala/org/apache/spark/sql/test/util/CarbonFunSuite.scala (hunk 2 above)
Comment (ShreelekhyaG) [reply integration/spark/src/main/scala/org/apache/spark/sql/test/util/CarbonFunSuite.scala:52]: done

File: integration/spark/src/main/scala/org/apache/carbondata/indexserver/IndexServer.scala (hunk 3 above)
Comment (ShreelekhyaG) [reply integration/spark/src/main/scala/org/apache/carbondata/indexserver/IndexServer.scala:54]: no need to ignore testcase as external carbon segments are pruned by indexserver. non-carbon segments are not pruned. From `distributedPrunedRdd `when async calls to prune are made, `TableIndex.prune` method throws IOException and this is not handled in parent caller - `indexserver.getsplits`

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_2.4.5/3736/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12602/job/ApacheCarbonPRBuilder2.3/5480/

**Comment by ShreelekhyaG** [comment]
2.3 build and UT with indexserver: http://121.244.95.60:12602/job/ApacheCarbon_Manual_2.3_Compile_and_UT/1/
Screenshot from 2021-06-03 18-22-02


2.4 build and UT with indexserver: http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_2.4_Compile_and_UT/65/
Screenshot from 2021-06-03 18-21-29

**Comment by Indhumathi27** [comment]
LGTM

**Comment by Indhumathi27** [comment]
retest this please

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12602/job/ApacheCarbonPRBuilder2.3/5495/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_2.4.5/3752/

**Comment by Indhumathi27** [comment]
retest this please

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_2.4.5/3754/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12602/job/ApacheCarbonPRBuilder2.3/5497/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_2.4.5/3756/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12602/job/ApacheCarbonPRBuilder2.3/5500/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_2.4.5/3761/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12602/job/ApacheCarbonPRBuilder2.3/5504/

**Comment by Indhumathi27** [comment]
retest this please

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12602/job/ApacheCarbon_PR_Builder_2.4.5/3769/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12602/job/ApacheCarbonPRBuilder2.3/5512/
